keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1107 @@
+    matchImages:
+      - "*.registry.io/*"
+    defaultCacheDuration: "10m"
+    apiVersion: credentialprovider.kubelet.k8s.io/v1
+    tokenAttributes:
+      serviceAccountTokenAudience: my-audience
+      # list of annotations that the plugin is interested in
+      # the annotation key and corresponding value in the service account will be passed to the plugin
+      serviceAccountAnnotationKeys:
+      - domain.io/identity-id
+      - domain.io/identity-type
+```
+
+##### Kubernetes API Server (KAS) Changes
+
+Add a new flag `--allowed-kubelet-audiences` to KAS to allow configuring the list of audiences for which the Kubelet is allowed to generate service account tokens for image pulls.
+
+The node audience restriction will be enforced by the [NodeRestriction admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction)
+and kubelet will only be allowed to generate a service account token for the audience configured in the credential provider configuration if it is present in the list of audiences configured in KAS.
+
+#### Credential Provider Request API
+
+Add `ServiceAccountToken` and `ServiceAccountAnnotations` fields to `CredentialProviderRequest`.
+
+1. `ServiceAccountToken` is the service account token bound to the pod for which the image is being pulled.
+    If the `ServiceAccountTokenAudience` field is configured in the kubelet's credential provider configuration, the token will be sent to the plugin.
+2. `ServiceAccountAnnotations` is a map of annotations on the KSA for which the image is being pulled. Only annotations defined in the `ServiceAccountAnnotations` field in the credential provider configuration will be passed to the plugin. If the `ServiceAccountAnnotations` field is not set in the configuration, this field will be empty.
+
+```diff
+--- a/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
++++ b/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
+@@ -32,6 +32,16 @@ type CredentialProviderRequest struct {
+        // credential provider plugin request. Plugins may optionally parse the image
+        // to extract any information required to fetch credentials.
+        Image string
++
++       // serviceAccountToken is the service account token bound to the pod for which
++       // the image is being pulled. This token is only sent to the plugin if the
++       // tokenAttributes.serviceAccountTokenAudience field is configured in the kubelet's credential provider configuration.
++       ServiceAccountToken string
++
++       // serviceAccountAnnotations is a map of annotations on the service account bound to the
++       // pod for which the image is being pulled. The list of annotations in the service account
++       // that need to be passed to the plugin is configured in the kubelet's credential provider
++       // configuration.
++       ServiceAccountAnnotations map[string]string
+ }
+```
+
+Example Kubernetes Service Account:
```

Comment:
Expand this example to show what happens when a credential provider requests an annotation key that is not set on the KSA.

Reply:
added a note about it in the credential provider request example.